### ISSUE TITLE
Make prediction loop parallel

### ISSUE THREAD

**Issue by djhocking** [Conte-Ecology/conteStreamTemperature_northeast#9]
Parallelize the for-loop (or while statement but might be harder) to make the daily predictions and derived metric calculations much faster.

The slow part is less the calculations and more about pulling the daymet records from Felek. I think this should speed it up but will there be any issues (or ping limitation) with calling to the SHEDS database on Felek repeatedly from the same account?

**Comment by djhocking** [comment]
Send to Denver cluster to make it more than 8X or 16X faster? Questions to ask:

Okay to pull data from external server repeatedly?
How to code the parallel loop for the server cluster?

**Comment by walkerjeffd** [comment]
I guess I would convert the for loop to an lapply and then use mclapply. This is what I use to parallelize code, and it typically works very well (though it can get tricky if one of the iterations fails). But you could probably use foreach or one of the other parallel packages too. I'm not sure if any one of them is better than the others.

The bigger challenge is pulling the data from the database. The trivial way to do that is to send one query for each featureid (have mclapply loop over all featureids, then for each featureid, fetch the data and compute the predictions). I guess one question is how long it takes to compute the predictions once the data has been fetched from the db. If it takes longer to fetch the data than to compute the predictions, then that may not work very well. 

You could also split it up into chunks, like HUC12. There are two new tables in the database: data.catchment_huc12 and data.catchment_state which have the relationship between featureid and huc12 and state (2 letter abbreviation). So you could first loop over HUC12s, then fetch all the data for the catchments within that huc12, then loop through the catchments in parallel and recombine at the end. One thing to note is that there are 163 catchments that are not associated with a HUC12 because the centroid is outside the HUC12 envelope.

From the database perspective, I don't think there is anything wrong with firing off many queries in succession. Just make sure you only set up one connection before the loop, then submit queries using that same connection. I think you'd get into trouble if it created a new connection for each iteration because there's a limiting maximum number of connections. I'd probably use RPostgreSQL and SQL queries directly instead of dplyr also.

**Comment by djhocking** [comment]
Okay, I've been working on this for a while and am still having trouble. Below is the code but you can find it at: https://github.com/Conte-Ecology/conteStreamTemperature_northeast/blob/master/predict_temperatures_parallel.R

When I run it the output log I create shows that it gets all the way to the last iteration before producing this error:

```
Error in { : 
  task 1 failed - "RS-DBI driver: (could not connect dan@ecosheds.org on dbname "sheds"
)"
In addition: Warning message:
In e$fun(obj, substitute(ex), parent.frame(), e$data) :
  already exporting variable(s): catchmentid, chunk.size, coef.list, cov.list, derive_metrics_par, df_covariates_upstream, featureid_huc8, featureid_lat_lon, logFile, n.loops, springFallBPs, tempDataSync, var.names
```

But it does it very fast, so I don't think it's actually doing the database pulls and calculations for each loop.

```
 library(foreach)
  #library(doMC) # doesn't work well with connecting to the database
  library(doParallel)
  library(RPostgreSQL)

  # size of chunks
  chunk.size <- 50
  n.loops <- ceiling(n.catches / chunk.size)

  # set up parallel backend & make database connection available to all workers
  nc <- min(c(detectCores()-1, 16)) # use the maximum number of cores minus 1 or up to 16 because max 16 database connections
  cl <- makePSOCKcluster(nc)
  registerDoParallel(cl)
  clusterEvalQ(cl, {
    library(DBI)
    library(RPostgreSQL)
    drv <- dbDriver("PostgreSQL")
    con <- dbConnect(drv, dbname='sheds', host='ecosheds.org', user=options('SHEDS_USERNAME'), password=options('SHEDS_PASSWORD'))
  })

  # setup to write out to monitor progress
  logFile = "localData/log_file.txt"
  cat("Monitoring progress of prediction loop in parallel", file=logFile, append=FALSE, sep = "\n")

  # start loop
  foreach(i = 1:n.loops, 
          .inorder=FALSE, 
          .combine = rbind,
          .packages=c("DBI", 
                      "RPostgreSQL",
                      "reshape2",
                      "dplyr",
                      "tidyr",
                      "zoo",
                      "lubridate",
                      "conteStreamTemperature"),
          .export=ls(envir=globalenv())
  ) %dopar% {

    # write start of each iteration
    #sink("localData/log.txt", append = TRUE)
    cat(paste("Starting iteration", i, " of ", n.loops, "\n"), file = logFile, append = TRUE)


    derived_metrics <- derive_metrics_par(i=i, chunk.size=chunk.size, catchmentid=catchmentid, springFallBPs=springFallBPs, tempDataSync=tempDataSync, df_covariates_upstream=df_covariates_upstream, featureid_huc8=featureid_huc8, featureid_lat_lon=featureid_lat_lon, coef.list=coef.list, cov.list=cov.list, var.names=var.names)

    #gc()
  }

  dbClearResult(res = rs)
  dbDisconnect(con)
  dbUnloadDriver(drv)
  dbListConnections(drv)

  gc()
```

**Comment by walkerjeffd** [comment]
I did some digging into this and I don't think you can initialize the connections for each worker in `clusterEvalQ()` before running the loop. When I tried it, I get a series of `<Expired PostgreSQLConnection:(26202,2)>` messages for each core. I think what you're trying to do is set up a `con` object for each worker, which is then used within `derive_metric_par()`, though I couldn't find where that function is defined and it doesn't look like you're passing a `con` object into it, so I'm not sure how you're getting the database connection into `derive_metric_par()`.

Also Section 2 of [Getting Started with doParallel and foreach](http://cran.r-project.org/web/packages/doParallel/vignettes/gettingstartedParallel.pdf) says that connection objects "very likely won't work". 

Another issue is that it doesn't look like you're shutting down the cluster when your done.  This section at the end probably doesn't do anything because `con` and `drv` are created within each worker:

``` r
dbClearResult(res = rs)
dbDisconnect(con)
dbUnloadDriver(drv)
dbListConnections(drv)
```

What you should do instead is run this after the `foreach` loop to close the connection within each worker, otherwise you'll keep creating new connections without ever closing them, which is probably how you ended up with 88 of them:

``` r
clusterEvalQ(cl, {
  dbDisconnect(con)
})
```

You also need to shut down the cluster otherwise you keep creating new R process forks whenever you run `cl <- makePSOCKcluster(nc)`. Chris and I noticed that at one point you have a few dozen R processes running. So make sure you run this after each `foreach`:

``` r
stopCluster(cl)
```

To find out how many R threads you have going, just SSH into felek and run:

``` sh
ps -u dan
```

and that will list all of your running processes. I'd suggest doing that before running any parallel code, then create the cluster (`cl <- makePSOCKcluster(nc)`) and you'll see that you have `nc` R processes running. Then run `stopCluster(cl)` and check your processes list and you'll see they disappeared. So use this to make sure you're closing R processes.

You can also monitor the number of PostgreSQL connections by running an SQL query within psql. So after ssh into felek, run:

``` sh
psql sheds
```

Then enter this SQL query which will list the number of connections for each user:

``` sql
select usename, count(*) from pg_stat_activity group by usename;
```

To quit psql, run `\q`.

But I think you can get your script to work by creating the db connection within each loop rather than before. I can't think of a reason why this would be more inefficient than trying to create one connection for each fork before running the loop. I tried this with a simpler query (not running `derive_metrics_par`) and it worked. Note that I close `con` after running `derive_metrics_par` but before returning from the loop. But as I said earlier, I'm not sure how `derive_metrics_par` is accessing the `con` object, so I think it needs to be argument to the function. 

``` r
cl <- makePSOCKcluster(nc)
registerDoParallel(cl)

# setup to write out to monitor progress
logFile = "localData/log_file.txt"
cat("Monitoring progress of prediction loop in parallel", file=logFile, append=FALSE, sep = "\n")

# start loop
foreach(i = 1:n.loops, 
        .inorder=FALSE, 
        .combine = rbind,
        .packages=c("DBI", 
                    "RPostgreSQL",
                    "reshape2",
                    "dplyr",
                    "tidyr",
                    "zoo",
                    "lubridate",
                    "conteStreamTemperature"),
        .export=ls(envir=globalenv())
) %dopar% {
  cat(paste("Starting iteration", i, " of ", n.loops, "\n"), file = logFile, append = TRUE)

  # create connection
  drv <- dbDriver("PostgreSQL")
  con <- dbConnect(drv, dbname='sheds', host='ecosheds.org', user=options('SHEDS_USERNAME'), password=options('SHEDS_PASSWORD'))

  derived_metrics <- derive_metrics_par(i=i, chunk.size=chunk.size, catchmentid=catchmentid, springFallBPs=springFallBPs, tempDataSync=tempDataSync, df_covariates_upstream=df_covariates_upstream, featureid_huc8=featureid_huc8, featureid_lat_lon=featureid_lat_lon, coef.list=coef.list, cov.list=cov.list, var.names=var.names)

  # close connection
  dbDisconnect(con)

  derived_metrics
}

stopCluster(cl)
```

You may also be having problems with `.export=ls(envir=globalenv()` in the foreach loop. I got mine to work even after commenting that out. I think that when the cluster is created, the processes are forked from the main process, which I think means they inherit the variables already, though I'm not sure. But you might try doing it without exporting the global environment.

**Comment by djhocking** [comment]
Thanks for all the help Jeff. It was working during my simple testing but now from what you've written I think I know where I went wrong. In my test, I wasn't running `derived_metrics_par` it was just the small chunk of the code that was doing the db pull. That is why `con` was working okay when in `clusterEvalQ()`. Then I think you're right, I was trying to close the connections but was not doing so within the workers, which is why they wouldn't close until they timed out. 

I was running `stopCluster(cl)` at the end, but I guess that didn't also close the connections, creating more problems. The `.export=ls(envir-globalenv())` call was in response to an error I was getting at one point, but I think it had to do with what was being passed to the function, not the workers. You're correct, I read that in most cases `doParallel` passes everything needed from the environment to the workers automatically although other parallel backend functions don't do that (doSnow?). I think I was also having some confusion because with the `parallel` package I had been using Forks rather than Sockets, so what gets copied for the workers is different.

The function is temporarily (while I've been modifying and testing) at the bottom of https://github.com/Conte-Ecology/conteStreamTemperature_northeast/blob/master/predict_temperatures_parallel.R

In the function I had

```
  # reconnect to database if lost
  if(isPostgresqlIdCurrent(con) == FALSE) {
    drv <- dbDriver("PostgreSQL")
    con <- dbConnect(drv, dbname='sheds', host='ecosheds.org', user=options('SHEDS_USERNAME'), password=options('SHEDS_PASSWORD'))
  }
```

Basically my strategy, which I realize was bad in retrospect, was to create a connection within each worker and leave it open while each worker went through it's loops but checking each time if the connection was lost and reconnecting in those cases. Then I would just close things all at the end, but of course that wasn't working because I was outside of the workers, so I was just accumulating connections every time I tried to run the loop.

Thanks for the help and the postgres tips. I will give it a try now.

**Comment by djhocking** [comment]
Works now quite quickly. Could transfer it back into a function in the R package later.
